src/Aspire.Hosting/Health/ResourceHealthCheckService.cs

```diff
@@ -13,21 +13,57 @@ namespace Aspire.Hosting.Health;
 
 internal class ResourceHealthCheckService(ILogger<ResourceHealthCheckService> logger, ResourceNotificationService resourceNotificationService, HealthCheckService healthCheckService, IServiceProvider services, IDistributedApplicationEventing eventing, TimeProvider timeProvider) : BackgroundService
 {
-    private readonly Dictionary<string, ResourceEvent> _latestEvents = new();
-    private readonly Dictionary<string, ResourceMonitorState> _resourcesStartedMonitoring = new();
+    private readonly Dictionary<string, ResourceMonitorState> _resourceMonitoringStates = new();
 
-    internal class ResourceMonitorState(CancellationToken serviceStoppingToken)
+    internal class ResourceMonitorState(ResourceEvent initialEvent, CancellationToken serviceStoppingToken)
     {
         private readonly CancellationTokenSource _cts = CancellationTokenSource.CreateLinkedTokenSource(serviceStoppingToken);
+        private TaskCompletionSource? _delayInterruptTcs;
+        private CancellationTokenSource? _delayCts;
 
         public CancellationToken CancellationToken => _cts.Token;
 
+        public ResourceEvent LatestEvent { get; private set; } = initialEvent;
+
         public void StopResourceMonitor()
         {
             _cts.Cancel();
         }
+
+        public void SetLatestEvent(ResourceEvent resourceEvent)
+        {
+            LatestEvent = resourceEvent;
+            _delayInterruptTcs?.TrySetResult();
```

Comment:
Do you want the delay to catch missed updates? SetLatestEvent and DelayAsync aren't synchronized. That means a call to SetLatestEvent  then Delay async won't yield.

I would suggest using a channel (instead of a tcs) if you don't want to miss updates. It would also remove allocations.

Reply:
Missing updates seems fine. There isn't a 1:1 relationship between the state of a resource and its health result, so timing between events will never be precisie. Everything is eventually consistent.

Reply:
I changed my mind. Previously a missed update when healthy would be caught by the polling in a second. Now there is always a 30 second wait until interrupted.

I added some locking and improved the interrupt decision to consider the resource version and state. I'm not 100% sure why a change in the snapshot should cause health checks to rerun, no tests cover that situation, but I think state change (such as `Running` to another state) is what it makes sense to consider.